### PR TITLE
Feature - Added a hook system into the baseline scripts & scans #4925

### PR DESCRIPTION
**Pull request by rezen**
As per the feature #4925 created a hook system to enable users to augment or add to the baseline/full/api scans. By default, the scripts will look for `~/.zap_hooks.py` but you can supply a `--hook=path-to-custom/hooks.py`

For each hook you want to plugin in to, you create a method with the same name as the hook and then there you can add to or change what zap is doing. Every hook that has the `zap_` prefix will pass the zap api object which you can interact with. Any hook defined in `zap_common.py` that uses the `@hook` decorator, you are intercepting the arguments that are passed to the method (the hook is the same name as the method) and you can modify the arguments. Every hook that ends with `_wrap` allows you to wrap the results/return value of the method of the same name.

**Example** 
```python
def start_docker_zap_wrap(container_id):
  return container_id

def zap_started(zap, target):
  print(zap)

def zap_access_target(zap, target):
  print("Access target %s" % target)

def zap_spider(zap, target):
  pass

def zap_get_alerts(zap, baseurl, blacklist, out_of_scope):
  blacklist.append('20001')
  return zap, baseurl, blacklist, out_of_scope

def zap_get_alerts_wrap(results):
  return results

def zap_spider_wrap(zap=None):
  print("Post", zap)
  raise Exception("test")

def zap_pre_shutdown(zap=None):
  pass
```

### CODE REVIEW

**Comment by psiinon** [comment]
This pull request **introduces 3 alerts** when merging a401aa12bfc02e2d37c9c522b07d4993c2eed19e into 300f5b403a6fd75cad6db5da42eea716ff0fa513 - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zaproxy/rev/pr-babbd5361290d8704f13d31c0e21f535fff5b7ca)

**new alerts:**

* 3 for Unused local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: docker/zap_common.py [hunk 1]
```diff
@@ -47,8 +48,86 @@
  the latest version.'''.replace('\n', '')
 
 zap_conf_lvls = ["PASS", "IGNORE", "INFO", "WARN", "FAIL"]
+zap_hooks = None
+
+def load_custom_hooks(hooks_file=None):
+    """ Loads a your custom python module which modifies zap scripts behaviour
```
Comment (psiinon) [inline docker/zap_common.py]: Either ' a ' or ' your ' not needed?

File: docker/tests/test_zap_hooks.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+    def zap_started(self, zap, target):
+        self.called += 1
+        return zap,
+
+
+class TestZapHooks(unittest.TestCase):
+
+    @contextmanager
+    def assertNotRaises(self, exc_type=Exception):
```
Comment (psiinon) [inline docker/tests/test_zap_hooks.py]: Raised instead of Raises?

File: docker/tests/test_zap_hooks.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    def test_trigger_hook_arg_length(self):
+        """ Triggering hook should return same args if no hook to trigger """
+        self.assertEqual(zap_common.zap_hooks, None)
+        args = [1, 2, 3, 4]
+        return_args = zap_common.trigger_hook('print_rules_wrap', *args)
+        self.assertEqual(len(args), len(return_args))
+
+
+    def test_trigger_hook_mismatch_no_exception(self):
+        """ If the hook signature doesn't match the hook, no exception is raised """
```
Comment (thc202) [inline docker/tests/test_zap_hooks.py]: Why not? This seems to evidence an error in the provided hook.

File: docker/tests/test_zap_hooks.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        zap_common.zap_hooks = _MockHooks()
+        with self.assertNotRaises():
+          zap_common.trigger_hook('zap_started')
+        self.assertEqual(zap_common.zap_hooks.called, 0)
+
+
+    def test_trigger_hook_ignore_exception(self):
+        """ Hooks that raise exceptions are ignored when triggered """
```
Comment (thc202) [inline docker/tests/test_zap_hooks.py]: This does not seem to be the best behaviour IMO, by continuing issues in the hooks could easily go unnoticed and the script would not be doing what the user expects/wants.

File: docker/tests/test_zap_hooks.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        with self.assertRaises(Exception):
+          zap_common.zap_hooks.bad_hook(1)
+        # Hook triggering goes on despite bad hook
+        with self.assertNotRaises():
+          zap_common.trigger_hook('bad_hook', 1)
+        self.assertEqual(zap_common.zap_hooks.called, 0)
+
+
+    def test_trigger_hook_verify_calls(self):
+        """ Should return same args if no hook to trigger """
```
Comment (thc202) [inline docker/tests/test_zap_hooks.py]: Does not seem to match the test.

File: docker/tests/test_zap_hooks.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        zap_common.trigger_hook('zap_started', *args)
+        zap_common.trigger_hook('zap_started', *args)
+        zap_common.trigger_hook('zap_started', *args)
+        zap_common.trigger_hook('zap_started', *args)
+        zap_common.trigger_hook('zap_started', *args)
+        self.assertEqual(zap_common.zap_hooks.called, 5)
+
+
+    def test_trigger_hook_maintain_signature(self):
+        """Should return same args if no hook to trigger """
```
Comment (thc202) [inline docker/tests/test_zap_hooks.py]: Same here.

File: docker/zap-api-scan.py [hunk 7]
```diff
@@ -148,12 +150,12 @@ def main(argv):
...
 
     try:
-        opts, args = getopt.getopt(argv, "t:f:c:u:g:m:n:r:J:w:x:l:hdaijp:sz:P:D:T:O:")
+        opts, args = getopt.getopt(argv, "t:f:c:u:g:m:n:r:J:w:x:l:hdaijp:sz:P:D:T:O:", ["hook="])
     except getopt.GetoptError as exc:
         logging.warning('Invalid option ' + exc.opt + ' : ' + exc.msg)
         usage()
         sys.exit(3)
-
+
```
Comment (thc202) [inline docker/zap-api-scan.py]: These trailing spaces should be removed (same in other places).

File: docker/zap-api-scan.py [hunk 8]
```diff
@@ -206,8 +208,13 @@ def main(argv):
             timeout = int(arg)
         elif opt == '-O':
             host_override = arg
+        elif opt == '--hook':
+            hook_file = arg
 
     check_zap_client_version()
+    
+    load_custom_hooks(hook_file)
+    opts = trigger_hook('cli_opts', opts)
```
Comment (thc202) [inline docker/zap-api-scan.py]: What is this one for? Or, does it worth assigning the return value even if `opts` is never used after?

File: docker/zap_common.py [hunk 9]
```diff
@@ -214,6 +294,7 @@ def wait_for_zap_start(zap, timeout_in_secs = 600):
           'Failed to connect to ZAP after {0} seconds'.format(timeout_in_secs))
 
 
+@hook(wrap=True)
```
Comment (thc202) [inline docker/zap_common.py]: What's the criteria being used to wrap?

File: docker/zap_common.py [hunk 10]
```diff
@@ -423,3 +509,12 @@ def write_report(file_path, report):
             report = report.encode('utf-8')
 
         f.write(report)
+
+@hook(wrap=True)
+def zap_import_context(zap, context_file):
+    # handle the context file, cant use base_dir as it might not have been set up
```
Comment (thc202) [inline docker/zap_common.py]: This should be removed, does not apply here.

**Comment by thc202** [comment]
The command line help is not enough to use this feature, the above description and example usage should be added to a wiki page and then linked from the existing script ones (e.g. https://github.com/zaproxy/zaproxy/wiki/ZAP-Baseline-Scan).

File: docker/tests/README.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# Testing
```
Comment (psiinon) [inline docker/tests/README.md]: Can we automatically run these tests in CI?

File: docker/tests/test_zap_hooks.py (hunk 2 above)
Comment (rezen) [reply docker/tests/test_zap_hooks.py]: I was trying to be consistent (and maybe illogical?) with the language of  `assertRaises` from stdlib but am not opposed to changing it.

**Comment by rezen** [comment]
@thc202 I wasn't sure how to update the wiki via a PR :/ should I just include a doc in this PR than can be copied into the wiki?

**Comment by thc202** [comment]
Yeah, the wiki does not support PRs, you can push directly though. Adding the doc here might be better, easier for review/feedback.

File: docker/zap_common.py (hunk 9 above)
Comment (thc202) [reply docker/zap_common.py]: Thanks.

File: docker/zap_common.py (hunk 10 above)
Comment (rezen) [reply docker/zap_common.py]: Why do you say that?

Comment (thc202) [reply docker/zap_common.py]: There's no `base_dir` here, the comment was from callers' scope.

Comment (rezen) [reply docker/zap_common.py]: Ah gotcha yes!

**Comment by rezen** [comment]
Addressed comments in last push :smile:

File: docker/docs/scan-hooks.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+hash (`http://example.com` vs `http://example.com/#/dashboard`). You can use the 
+`zap_ajax_spider` to intercept the arguments and modify them to what you want.
+
+**Running Scripts**  
+You may have some standalone scripts that you want to run before crawling & the 
+other parts of the scan. You can use the hook `zap_started` to run some scripts 
+before the rest of the scan runs.
+
+**Changing Configs**  
+For some applications, there are a number active scan scanners that you may want
```
Comment (thc202) [inline docker/docs/scan-hooks.md]: `number of`?

File: docker/docs/scan-hooks.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+disable that are not applicable for that application. Configuring your policy
+before the active scan using `zap_active_scan` hook can ensure you only run the 
+tests you want to run.
+
+
+## Example
+#### Define your hooks in a python file `my-hooks.py`
+You define all the hooks you want to integrate with with python methods that
+correspond with the name of the hook. By default, zap scans will look for
```
Comment (thc202) [inline docker/docs/scan-hooks.md]: `ZAP`.

File: docker/tests/test_zap_hooks.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    def test_trigger_hook_mismatch_exception(self):
+        """ If the hook signature doesn't match the hook the exception bubbles up """
+        zap_common.zap_hooks = _MockHooks()
+        with self.assertRaises(Exception):
+          zap_common.trigger_hook('zap_started')
+        self.assertEqual(zap_common.zap_hooks.called, 0)
+
+
+    def test_trigger_hook_verify_calls(self):
+        """ Verify the hook gets called if it matchesShould return same args if no hook to triggerr signature """
```
Comment (thc202) [inline docker/tests/test_zap_hooks.py]: This needs tweaks.

File: docker/tests/test_zap_hooks.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        zap_common.zap_hooks = _MockHooks()
+        args = ['zap', 'http://127.0.0.1']
+        zap_common.trigger_hook('zap_started', *args)Should return same args if no hook to trigger
```
Comment (thc202) [inline docker/tests/test_zap_hooks.py]: Here (and following) as well?

File: docker/tests/test_zap_hooks.py (hunk 6 above)
Comment (thc202) [reply docker/tests/test_zap_hooks.py]: Bump, the description does not match the test (I think).

**Comment by rezen** [comment]
Reread the docs, fixed some additional grammatical errors. For the tests there was a haphazard paste that multiplied itself :disappointed: fixed that!

**Comment by thc202** [comment]
Thank you!

**Comment by rezen** [comment]
If you like I can do a blog post on using this feature on my site?

**Comment by thc202** [comment]
Sounds great! It could also be published in ZAP's blog?

**Comment by psiinon** [comment]
@rezen agreed, would be very happy for you to publish this on the ZAP blog :)

**Comment by rezen** [comment]
Ya definitely please use the content! I'm a couple weeks weeks out from getting to it. I'll also use some examples to address some users' issues. The hook system is great for addressing one off features/requests to scripts :smile: 

- https://github.com/zaproxy/zaproxy/issues/3809
- https://github.com/zaproxy/zaproxy/issues/4354

**Pull request merged**